src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/CSharpTagHelperCodeRenderer.cs

```diff
@@ -230,48 +230,67 @@ private void RenderTagHelpersCreation(TagHelperChunk chunk)
+                        // helpers that need the value.
                         htmlAttributeValues.Add(attributeDescriptor.Name, valueAccessor);
 
                         if (bufferableAttribute)
                         {
-                            // If the attribute is bufferable but has a plain text value that means the value
-                            // is a string which needs to be surrounded in quotes.
+                            _writer.WriteStartAssignment(valueAccessor);
+
                             if (isPlainTextValue)
                             {
+                                // If the attribute is bufferable but has a plain text value that means the value
+                                // is a string which needs to be surrounded in quotes.
                                 RenderQuotedAttributeValue(textValue, attributeDescriptor);
                             }
                             else
                             {
-                                // The value contains more than plain text. e.g. someAttribute="Time: @DateTime.Now"
+                                // The value contains more than plain text e.g.
+                                // stringAttribute ="Time: @DateTime.Now"
                                 RenderBufferedAttributeValue(attributeDescriptor);
                             }
+
+                            _writer.WriteLine(";");
                         }
                         else
                         {
-                            // TODO: Make complex types in non-bufferable attributes work in
-                            // https://github.com/aspnet/Razor/issues/129
-                            if (!isPlainTextValue)
+                            // Write out simple assignment for non-string property value. Try to keep the whole
+                            // statement together and the #line pragma correct to make debugging possible.
+                            using (var lineMapper = new CSharpLineMappingWriter(
+                                _writer,
+                                attributeValueChunk.Association.Start,
+                                _context.SourceFile))
                             {
+                                // Place the assignment LHS to align RHS with original attribute value's indentation.
+                                // Unfortunately originalIndent is incorrect if original line contains tabs. Unable to
+                                // use a CSharpPaddingBuilder because the Association has no Previous node; lost the
+                                // original Span sequence when the parse tree was rewritten.
+                                var originalIndent = attributeValueChunk.Start.CharacterIndex;
+                                var generatedLength = valueAccessor.Length + " = ".Length;
+                                var newIndent = originalIndent - generatedLength;
```

Comment:
@NTaylorMullen @pranavkm and I agree the best-effort code I have here mimics what he did in `RenderRuntimeExpressionBlockChunkWithContentSpan`.  If we want to centre things, that would be a separate bug and it would require calculation of where the statement or expression ends.